client/master_client.go

```diff
@@ -24,25 +25,51 @@ type clientHolder struct {
 }
 
 type MasterClient struct {
-	urls    []string
-	leader  string
-	clients map[string]*clientHolder
+	urls        []string
+	leader      string
+	clientsLock sync.RWMutex
+	clients     map[string]*clientHolder
+}
+
+func (c *MasterClient) dialMaster(ctx context.Context, addr string) error {
+	ctx, cancel := context.WithTimeout(ctx, dialTimeout)
+	defer cancel()
+	conn, err := grpc.DialContext(ctx, addr, grpc.WithInsecure(), grpc.WithBlock())
+	if err != nil {
+		return errors.Wrap(errors.ErrGrpcBuildConn, err)
+	}
+	c.clients[addr] = &clientHolder{
+		conn:   conn,
+		client: pb.NewMasterClient(conn),
+	}
+	return nil
+}
+
+// UpdateClients receives a list of server master addresses, dials to server
+// master that is not maintained in current MasterClient.
+func (c *MasterClient) UpdateClients(ctx context.Context, urls []string) {
+	for _, addr := range urls {
+		// TODO: refine address with and without scheme
+		addr = strings.Replace(addr, "http://", "", 1)
+		if _, ok := c.clients[addr]; !ok {
+			c.urls = append(c.urls, addr)
+			log.L().Info("add new server master client", zap.String("addr", addr))
+			err := c.dialMaster(ctx, addr)
+			if err != nil {
```

Comment:
if error , still store the address in the c.urls?

Reply:
Good catch, added some refine code and unit test.